### ISSUE TITLE
Question: what would be the reason for jobs not being processed but moved to completed state?

### ISSUE THREAD

**Issue by PrasannaaSanthanam** [OptimalBits/bull#1530]
Hi , 

We are using bull to do some async jobs. We create the jobs and & push them to respective Queues (there are 3 different Queues for 3 different types of jobs).

At times, we notice that, some of jobs once pushed to the Queue, enter waiting state, but never reaches 'active' state, so those jobs never gets processed. But we when we get the details of those particular jobs, they are in completed state. **There are no events when the job gets moved to completed state as well.**

This happens randomly but frequently, we cannot see a consistent pattern in this issue. 
For ex, If we push 10 jobs to the Queue, chances are that some of the jobs are successfully completed, wheres as some are moved to completed without processing.

Notes:
* 3 different Queues created with unique names.
* Jobs are not named
* No concurrency set on process method
* We do reuse Redis connections as mentioned in (https://github.com/OptimalBits/bull/blob/develop/PATTERNS.md#reusing-redis-connections)
* Redis is not not clustered.

Let me know if you need more info.

Versions:
Bull Version: 3.11.0
Node Js: 10.15.3

**Comment by manast** [comment]
jobs needs to be processed before they can be moved to completed, please try to create a simple code snippet that reproduces the issue and we can look into it in more detail.

**Comment by PrasannaaSanthanam** [comment]
Hi @manast,
Thanks for the reply, I tried to repo this using a simple code snippet. Unfortunately the issue does not happen.. Also I noticed a few things after my last post.

I recently changed all my queue names and then the jobs were processed properly for about 3 days. Now the same issue has started happening. I see the job in waiting state and never reaches "Active" state, there by my process callback never gets executed, but If I get the job details, I can see that both "finishedOn" & "processedOn" has timestamps and there is not a lot of difference between these two timestamps. 

I have logged all the life cycle events on the queue. So, I should be seeing the logs for this job when it becomes active and as well as complete.. But, I never see them on my logs.

* Are there any limits on the jobs that a Queue can handle? Because, with a fresh queue this issue does not happen, only after processing for a lot of jobs I can see this behavior.

* Our Redis is on AWS (non-clustered) setup as master slave.. are there any recommendations for this setup w.r.t to Jobs scheduling?

* No reuse of redis clients across queues. 

* Are there any ways I can debug this issue? As of now I only have the event life cycle logs, which does not help much.

Any help is appreciated.

Job details **(which never became active and processed)**
{
    "id": "1675",
    "data": {
       ....
    },
    "opts": {
        "delay": 0,
        "attempts": 5,
        "stackTraceLimit": 10,
        "backoff": {
            "type": "exponential",
            "delay": 10000
        },
        "removeOnComplete": 100,
        "timestamp": 1573485835218
    },
    "progress": 0,
    "delay": 0,
    "timestamp": 1573485835218,
    "attemptsMade": 0,
    "stacktrace": [],
    "returnvalue": null,
    **"finishedOn": 1573485835392,
    "processedOn": 1573485835220**
}

**Comment by manast** [comment]
Here some answers:

> Are there any limits on the jobs that a Queue can handle? Because, with a fresh queue this issue does not happen, only after processing for a lot of jobs I can see this behavior.

There are no limits other than the size of your redis instance.

> Our Redis is on AWS (non-clustered) setup as master slave (v 3.x).. are there any recommendations for this setup w.r.t to Jobs scheduling?

If you mean running as Elasticache then there are no known issues. Non-clustered setup also, no known issues.

> Will the jobs gets evicted from redis, when jobs are put in waiting state for a long time?

No eviction.

> Are there any ways I can debug this issue? As of now I only have the event life cycle logs, which does not help much.

you can use REDIS MONITOR command, Redis commander to study the structures in Redis or https://taskforce.sh

> Also, will there be an issue if there are multiple jobs in the queue with same data info(job related info), In my case, this is a common scenario, where the user might trigger the same job multiple times. Will the queue in anyway think/assume that the job is already processed since the data node is same for all these jobs and move them to completed state without going through "Active" state?

No as long as they have different jobIds.

**Comment by PrasannaaSanthanam** [comment]
Hi @manast 

Thanks for the reply again... I noticed one more thing... after my last post. So far I was listening only to local events and not for global events. After my last post I also started listening for global events. With global event listener, I notice that jobs which are processed successfully has event logs from both local and global event listener callbacks. But the jobs which are not actually processed by my callback emits only the global event logs and not the local event logs, which i think explains why that specific job is not at all processed..

I am attaching the event logs for 2 consecutive jobs here... the first job(1718) is successfully processed and both local and global events are seen from both the nodes, where the next job(1719) is not processed and **I can see ONLY global events but not local events**...

NODE A
[INFO][2019-11-11 18:33:02][ip-10-4-1-145] [-] - onActiveEvent {"jobId":"1718","attempts":0}
[INFO][2019-11-11 18:33:02][ip-10-4-1-145] [-] - onWaitingEvent {"jobId":"1718"}        // local
[INFO][2019-11-11 18:33:02][ip-10-4-1-145] [-] - onGlobalWaitingEvent {"jobId":"1718"}
[INFO][2019-11-11 18:33:02][ip-10-4-1-145] [-] - onGlobalActiveEvent {"jobId":"1718"}
[INFO][2019-11-11 18:33:03][ip-10-4-1-145] [-] - onCompleteEvent {"jobId":"1718","attempts":0} // local
[INFO][2019-11-11 18:33:03][ip-10-4-1-145] [-] - onGlobalCompleteEvent {"jobId":"1718"}
[INFO][2019-11-11 18:33:25][ip-10-4-1-145] [-] - onWaitingEvent {"jobId":"1719"} // local
[INFO][2019-11-11 18:33:25][ip-10-4-1-145] [-] - onGlobalWaitingEvent {"jobId":"1719"}
[INFO][2019-11-11 18:33:25][ip-10-4-1-145] [-] - onGlobalActiveEvent {"jobId":"1719"}
[INFO][2019-11-11 18:33:26][ip-10-4-1-145] [-] - onGlobalCompleteEvent {"jobId":"1719"}


NODE B
[INFO][2019-11-11 18:33:02][ip-10-4-2-163] [-] - onGlobalWaitingEvent {"jobId":"1718"}
[INFO][2019-11-11 18:33:02][ip-10-4-2-163] [-] - onGlobalActiveEvent {"jobId":"1718"}
[INFO][2019-11-11 18:33:03][ip-10-4-2-163] [-] - onGlobalCompleteEvent {"jobId":"1718"}
[INFO][2019-11-11 18:33:25][ip-10-4-2-163] [-] - onGlobalWaitingEvent {"jobId":"1719"}
[INFO][2019-11-11 18:33:25][ip-10-4-2-163] [-] - onGlobalActiveEvent {"jobId":"1719"}
[INFO][2019-11-11 18:33:26][ip-10-4-2-163] [-] - onGlobalCompleteEvent {"jobId":"1719"}

Does this shed any light on the issue?

**Comment by manast** [comment]
A local event can only trigger on the worker that actually processed the job, so in this case in NODE A.

**Comment by PrasannaaSanthanam** [comment]
For jobId 1719, I do not see any local events on both Nodes, but I can see global events... So, without no  local events, can a global event fire?

**Comment by manast** [comment]
maybe you have other workers running somewhere else "stealing" jobs from the queue.

**Comment by PrasannaaSanthanam** [comment]
Hi @manast, thanks for the hint (stealing)... Yes, the jobs were consumed by another lower service, since redis is shared across lower envs. So as a fix, I have prefixed the queue names along with their envNames so that the queues are unique across shared redis. This way the jobs will be consumed only with in the env nodes and across envs.

I sincerely appreciate your patience and for all the replies... 
I am not sure if it helps others if we mention to have unique name across env when redis is shared.I did read that jobs has to have unique names, but having a shared redis slipped my mind.

I will close this issue now.. Thanks again.
